ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/TopicStream.java

```diff
@@ -52,96 +60,196 @@ private TopicStream() {
+      );
+    }
 
-    public RecordFormatter(final SchemaRegistryClient schemaRegistryClient,
-                           final String topicName) {
-      this.topicName = Objects.requireNonNull(topicName, "topicName");
+    @VisibleForTesting
+    RecordFormatter(
+        final SchemaRegistryClient schemaRegistryClient,
+        final String topicName,
+        final DateFormat dateFormat
+    ) {
+      this.topicName = requireNonNull(topicName, "topicName");
       this.avroDeserializer = new KafkaAvroDeserializer(schemaRegistryClient);
+      this.dateFormat = requireNonNull(dateFormat, "dateFormat");
     }
 
-    public List<String> format(final ConsumerRecords<String, Bytes> records) {
-      return StreamSupport
-          .stream(records.records(topicName).spliterator(), false)
-          .filter(Objects::nonNull)
-          .filter(r -> r.value() != null)
-          .filter(r -> r.value().get() != null)
-          .filter(r -> r.value().get().length != 0)
-          .map((record) -> {
-            if (formatter == null) {
-              formatter = getFormatter(record);
-            }
-            try {
-              return formatter.print(record);
-            } catch (IOException e) {
-              log.warn("Exception formatting record", e);
-              return null;
-            }
-          })
-          .filter(Objects::nonNull)
+    public List<Supplier<String>> format(final Iterable<ConsumerRecord<Bytes, Bytes>> records) {
+      if (!keyFormatter.isPresent()) {
+        keyFormatter = getKeyFormatter(records);
+      }
+
+      if (!valueFormatter.isPresent()) {
+        valueFormatter = getValueFormatter(records);
+      }
+
+      return StreamSupport.stream(records.spliterator(), false)
+          .map(this::delayedFormat)
           .collect(Collectors.toList());
     }
 
-    public Format getFormat() {
-      return formatter == null ? Format.UNDEFINED : formatter.getFormat();
+    @SuppressWarnings("OptionalGetWithoutIsPresent") // will not be empty if needed
+    private Supplier<String> delayedFormat(final ConsumerRecord<Bytes, Bytes> record) {
+      return () -> {
+        try {
+          final String rowTime = record.timestamp() == ConsumerRecord.NO_TIMESTAMP
+              ? "N/A"
+              : dateFormat.format(new Date(record.timestamp()));
+
+          final String rowKey = record.key() == null || record.key().get() == null
+              ? "<null>"
+              : keyFormatter.get().print(record.key());
+
+          final String value = record.value() == null || record.value().get() == null
+              ? "<null>"
+              : valueFormatter.get().print(record.value());
+
+          return "rowtime: " + rowTime
+              + ", " + "key: " + rowKey
+              + ", value: " + value;
+        } catch (IOException e) {
+          log.warn("Exception formatting record", e);
+          return "Failed to parse row";
+        }
+      };
+    }
+
+    public String getKeyFormat() {
```

Comment:
nit: `getKeyFormatName`

Reply:
Maybe. Though what's being returned hasn't changed really. Sure it returned an enum before, but the only place that was used changed it to a string.  Now it returns a string containing the details of the format.